packages/connection-manager/src/ConnectionQueue.js

```diff
@@ -0,0 +1,112 @@
+import { List } from "immutable";
+import ConnectionQueueItem from "./ConnectionQueueItem";
+
+/**
+ * Connection Queue
+ *
+ * The queue is prioritized to enable dynamic connection handling and
+ * best utilize the bandwidth available.
+ *
+ * This impl. wraps `Immutable.List` instead of extending `List` due to
+ * https://github.com/facebook/immutable-js/issues/301
+ */
+export default class ConnectionQueue {
+  /**
+   * Initializes an instance of ConnectionQueue
+   * @param {Set} connections – list of connections
+   */
+  constructor(connections = List()) {
+    if (!List.isList(connections)) {
+      throw new Error(
+        "Given list of connection must to be an instance of List."
+      );
+    }
+
+    /**
+     * @property {List}
+     * @name ConnectionQueue#connections
+     */
+    Object.defineProperty(this, "connections", { value: connections });
+  }
+
+  /**
+   * Return current size of queue
+   * @return {int} size of queue
+   */
+  get size() {
+    return this.connections.size;
+  }
+
+  /**
+   * returns the first connection
+   * @return {AbstractConnection} - first connection
+   */
+  first() {
+    if (this.connections.first() === undefined) {
+      return undefined;
+    }
+
+    return this.connections.first().connection;
+  }
+
+  /**
+   * returns a new queue without the first connection
+   * @return {ConnectionQueue} – ConnectionQueue without first connection
+   */
+  shift() {
+    return new ConnectionQueue(this.connections.shift());
+  }
+
+  /**
+   * Adds given connection to queue
+   * @param {AbstractConnection} connection – Connection to be added to queue
+   * @param {Integer} [priority] – priority of connection
+   * @return {ConnectionQueue} - ConnectionQueue containing the added connection
+   */
+  enqueue(connection, priority) {
+    const item = new ConnectionQueueItem(connection, priority);
+    const index = this.connections.findIndex(arrayItem =>
+      arrayItem.equals(item)
+    );
+    let connections = null;
+
+    if (index !== -1) {
+      connections = this.connections.update(index, item);
```

Comment:
Do we really want to update if `ConnectionQueueItem`s are the same, meaning they hold the same connection?

Reply:
yup, because the priority might have changed and we cant check it. (or in other words: checking for this would be as expensive as just updating)